### PR TITLE
React components cannot render undefined

### PR DESCRIPTION
**Pull request by jacobwgillespie**
React does not allow components to render `undefined`, only `null` is permitted.  Returning `undefined` from the render method results in the error:

> **Invariant Violation**
> ComponentName(...): Nothing was returned from render. This usually means a return statement is missing. Or, to render nothing, return null.

See also the discussion at https://github.com/facebook/react/issues/3738.

This PR is a continuation of #25546 by @elongl.

---

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/react/issues/3738
- [x] Increase the version number in the header if appropriate. **(n/a)**
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. **(n/a)**

### CODE REVIEW

**Comment by ferdaber** [comment]
`ReactNode` is not the same as the allowable return types of JSX element factories, unfortunately, due to how the compiler checks compatibility of JSX components. `ReactNode` is... problematic but we need `undefined` to be kept there for the time being to be able to render `undefined` children, which *are* valid.

**Comment by Kovensky** [comment]
I'm actually surprised at how few changes are needed to fix this part of `ReactNode`'s problems 🤔

I think removing `undefined` from it is fine here. `undefined` is not allowed as the result type of either function components or class components' `render()`.

We can't fix function components properly because of the compiler being weird but we can fix class components.

File: types/react/index.d.ts [hunk 1]
```diff
@@ -169,7 +169,7 @@ declare namespace React {
 
     interface ReactNodeArray extends Array<ReactNode> {}
     type ReactFragment = {} | ReactNodeArray;
-    type ReactNode = ReactChild | ReactFragment | ReactPortal | boolean | null | undefined;
+    type ReactNode = ReactChild | ReactFragment | ReactPortal | boolean | null;
```
Comment (Kovensky) [inline types/react/index.d.ts]: If you're removing `undefined` from `ReactNode` you need to re-introduce it in `ReactNodeArray` (`extends Array<ReactNode | undefined>`) as `undefined` inside an array is valid.

Comment (jacobwgillespie) [reply types/react/index.d.ts]: Good catch, done!

**Review by huan086 — APPROVED** [review]
Approved for changes in `types/lingui__macro/ReactSelect.d.ts` and `types/lingui__react/Select.d.ts`

**Comment by huan086** [comment]
@jacobwgillespie The `ReactNode: ... | undefined` is introduced in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/11935. I don't quite understand the changes though.

The pull request specifically mentioned `React.cloneElement`. Please check if it allows undefined.

Please check all the definitions that involves `children: ReactNode[]`, because the fix for `ReactNodeArray` might not be enough.

**Comment by Kovensky** [comment]
(an aside: I think ReactFragment should be the correct type name instead of ReactNodeArray (it's a "keyed fragment" in reconciler terms), but we can't touch that one yet because of the {})

**Comment by ericanderson** [comment]
We are comfortable bumping to TS 3.x. We just need someone to put in the pr

**Comment by johnnyreilly** [comment]
Is this ready to merge?

**Review by huan086 — CHANGES_REQUESTED** [review]
Pending changes to `ReactNode[]`

**Comment by jacobwgillespie** [comment]
Okay, I've updated the PR with the above changes:

* If something expects an array of React children, like the `children` arg to `createElement`, it uses `ReactNodeArray`
* If something typed `children` as `ReactNode | ReactNode[]`, it's now just `ReactNode` since that already includes `ReactNodeArray`
* The React types now require TypeScript 3.1
* Any types that depend on React now require TypeScript 3.1

I broke the changes out into individual commits, to assist with review.

**Comment by jacobwgillespie** [comment]
Looks like the `// TypeScript Version: x.x` comment isn't preferred anymore, let me update the `typesVerisons`...

**Comment by jacobwgillespie** [comment]
@ericanderson / @johnnyreilly can you advise on how to proceed, given the above?  I'm not sure the proper procedure for migrating to `typesVersions`.

**Comment by johnnyreilly** [comment]
Hey @jacobwgillespie ,

I'm afraid I haven't played with `typesVersions` myself yet and so I know only as much as you; apologies.

**Comment by nasreddineskandrani** [comment]
removing `undefined`... `null` kept...

https://basarat.gitbooks.io/typescript/docs/javascript/null-undefined.html
> Final thoughts
TypeScript team doesn't use null : TypeScript coding guidelines and it hasn't caused any problems. Douglas Crockford thinks null is a bad idea and we should all just use undefined.

https://github.com/Microsoft/TypeScript/issues/7426#issuecomment-228630489

**Comment by Jessidhia** [comment]
The null-undefined page you linked is but a style guide, telling you which one to choose in your own code. DefinitelyTyped, however, represents _other people's_ code, and sometimes code _does_ make a distinction between `null` and `undefined` (like React, which we wanted to fix here), and when that happens we also have to represent it here.

**Comment by nasreddineskandrani** [comment]
> sometimes code does make a distinction between null and undefined (like React,
@Jessidhia exact it's why i can't avoid `null`. I guess it's just better to give up and use `==` and `!=`. 
The problem that is killing me is when you arrive in a project in a company and devs do use the 2 with the same intent :(. With `stictNullChecks` you never know which one to expect... some devs really crack the code and just push feature without thinking twice... and it's just impossible to review all PRs in large teams... I wanted to auto pick `undefned` and it cannot be....

**Pull request closed without merging**